apps/pipelines/nodes/nodes.py

```diff
@@ -488,3 +507,55 @@ def post_extraction_hook(self, output, state):
                 team=session.team,
                 data=output,
             )
+
+
+class AssistantNode(Passthrough):
+    __human_name__ = "OpenAI Assistant"
+    __node_description__ = "Calls an OpenAI assistant"
+    assistant_id: AssistantId
+    citations_enabled: ToggleField = Field(default=True, description="Whether to include cited sources in responses")
+    input_formatter: str | None = Field(description="(Optional) Use {input} to designate the user input")
+
+    @field_validator("input_formatter")
+    def ensure_input_variable_exists(cls, value):
+        value = value or ""
+        acceptable_var = "input"
+        prompt_variables = set(PromptTemplate.from_template(value).input_variables)
+        if value:
+            if acceptable_var not in prompt_variables:
+                raise PydanticCustomError("invalid_input_formatter", "The input formatter must contain {input}")
+
+            acceptable_vars = set([acceptable_var])
+            extra_vars = prompt_variables - acceptable_vars
+            if extra_vars:
+                raise PydanticCustomError("invalid_input_formatter", "Only {input} is allowed")
+
+    def _process(self, input, state: PipelineState, node_id: str, **kwargs) -> str:
+        assistant = OpenAiAssistant.objects.get(id=self.assistant_id)
+        runnable = self._get_assistant_runnable(assistant, session=state["experiment_session"])
+        attachments = [Attachment.model_validate(params) for params in state.get("attachments", [])]
+        chain_output: ChainOutput = runnable.invoke(input, config={}, attachments=attachments)
+        output = chain_output.output
+
+        return PipelineState(
+            messages=[output],
+            outputs={node_id: output},
+            message_metadata={
+                "input": runnable.state.get_message_metadata(ChatMessageType.HUMAN),
+                "output": runnable.state.get_message_metadata(ChatMessageType.AI),
```

Comment:
Seeing this now I think in the long run it would be better to have this added to the chain output. OK to leave it for now though.

Reply:
Cool, agreed. We can then also remove [this ugly](https://github.com/dimagi/open-chat-studio/blob/7c2a37cb157a6e71f683e954260f2a2f772c3f0e/apps/chat/bots.py#L227) code by returning a proper object + metadata.